packages/drivers/appium/index.ts

```diff
@@ -51,6 +56,43 @@ export class WebdriverIOAppiumFrameworkDriver
     }
   }
 
+  /** Waits till the view hierarchy is stable */
+  private async waitForStableState(
+    pollInterval: number = DEFAULT_POLL_INTERVAL,
+    timeout: number = DEFAULT_TIMEOUT,
+  ): Promise<string | undefined> {
+    const startTime = Date.now();
+    let lastSnapshot: string | undefined;
+
+    while (Date.now() - startTime < timeout) {
+      const currentSnapshot: string = (await driver.getPageSource()) || "";
+      if (!currentSnapshot) {
+        return undefined;
+      }
+
+      if (lastSnapshot) {
+        const isStable = this.compareViewHierarchies(
+          currentSnapshot,
+          lastSnapshot,
+        );
+        if (isStable) {
+          return currentSnapshot;
+        }
+      }
+
+      lastSnapshot = currentSnapshot;
+      await new Promise((resolve) => setTimeout(resolve, pollInterval));
+    }
+
+    // Return the last snapshot if timeout is reached
+    return lastSnapshot;
+  }
+  private compareViewHierarchies(current: string, last: string): boolean {
+    // Use MD5 for fast comparison of view hierarchies
+    const currentHash = crypto.createHash("md5").update(current).digest("hex");
+    const lastHash = crypto.createHash("md5").update(last).digest("hex");
+    return currentHash === lastHash;
+  }
```

Comment:
extract to a util (e.g. `waitForStableViewHierarchy.ts`)

Reply:
maybe `getStableViewHierarchy ` will be clearer here, since we're returning the hierarchy when it's stable